### PR TITLE
Fix `pytest.mark.parametrize` when the argvalue is an iterator

### LINKED ISSUE #5354: Version 4.6.0 skips tests without apparent reason
Since version 4.6.0 pytest skips tests without apparent reason: https://travis-ci.org/Snawoot/postfix-mta-sts-resolver/jobs/540181138

- [x] output of `pip list` from the virtual environment you are using: **[HERE](https://travis-ci.org/Snawoot/postfix-mta-sts-resolver/jobs/540181138#L476)**
- [x] pytest and operating system versions: **pytest 4.6.0 on Ubuntu Xenial @ Travis CI**
- [x] minimal example if possible: **link above**

I can't understand why it happens, so I had to immediately rollback to 4.5.0 and fix this version in dev dependencies.

### PR DESCRIPTION
**Pull request by asottile**
Resolves #5354

### CODE REVIEW

**Comment by Sup3rGeo** [comment]
Thanks @asottile ! If I understand correctly, we were consuming the generator when checking for parametrization arguments to ignore at first so when the real parametrization kicked in there was nothing left for the parameters, right?

**Comment by asottile** [comment]
> Thanks @asottile ! If I understand correctly, we were consuming the generator when checking for parametrization arguments to ignore at first so when the real parametrization kicked in there was nothing left for the parameters, right?

that is exactly right :D

File: testing/test_mark.py [hunk 1]
```diff
@@ -413,6 +413,27 @@ def test_func(a, b):
     assert result.ret == 0
 
 
+def test_parametrize_iterator(testdir):
```
Comment (nicoddemus) [inline testing/test_mark.py]:
```suggestion
def test_parametrize_iterator(testdir):
    """parametrize should work with generators (#5354)."""
```

**Comment by asottile** [comment]
#5357

The procedure I used here was:

```bash
git checkout origin/4.6-maintenance -b backport_5356
git cherry-pick ...
```

**Comment by blueyed** [comment]
Removed "needs backport" label - I think we should do this after the backport was done, to track closed PRs that still need one.
Agreed?

**Comment by RonnyPfannschmidt** [comment]
this pr doenst test critical cases in, and it should fail on iterators

File: testing/test_mark.py [hunk 2]
```diff
@@ -413,6 +413,28 @@ def test_func(a, b):
...
+    py_file = testdir.makepyfile(
+        """\
+        import pytest
+
+        def gen():
+            yield 1
+            yield 2
+            yield 3
+
+        @pytest.mark.parametrize('a', gen())
```
Comment (RonnyPfannschmidt) [inline testing/test_mark.py]: this breaks down on marker/generator reuse

**Comment by nicoddemus** [comment]
> Removed "needs backport" label - I think we should do this after the backport was done, to track closed PRs that still need one.
Agreed?

Agreed, I had the same idea yesterday. 👍

**Comment by asottile** [comment]
Let's not lose track of "did we backport this?" let's replace the label when competed with a backported label

**Comment by blueyed** [comment]
Ok.
Although there is / should typically be a referencing comment anyway.

We should/could also do backports in batches then - resulting in a single CI run, instead of one for every backport.

File: testing/test_mark.py (hunk 2 above)
Comment (blueyed) [reply testing/test_mark.py]: @RonnyPfannschmidt please consider creating a new issue - or do you know what's up with the comment, @asottile ?

Comment (asottile) [reply testing/test_mark.py]: I think @RonnyPfannschmidt is suggesting this won't work (not at a computer)

```python
m = pytest.mark.parametrize

@m 
def test1(): pass

@m 
def test2(): pass
```

Comment (asottile) [reply testing/test_mark.py]: But if it doesn't, it never worked so :man_shrugging: probably?

Comment (RonnyPfannschmidt) [reply testing/test_mark.py]: we already had a issue about reuse of generators

Comment (asottile) [reply testing/test_mark.py]: yeah that example above is indeed broken, but wasn't the regression being targetted by this fix:

```console
$ pytest t.py -vv
============================= test session starts ==============================
platform linux -- Python 3.6.7, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /tmp/venv/bin/python3
cachedir: .pytest_cache
rootdir: /tmp
plugins: celery-4.3.0
collected 4 items                                                              

t.py::test1[1] PASSED                                                    [ 25%]
t.py::test1[2] PASSED                                                    [ 50%]
t.py::test1[3] PASSED                                                    [ 75%]
t.py::test2[a0] SKIPPED                                                  [100%]

===================== 3 passed, 1 skipped in 0.01 seconds ======================
```

**Comment by nicoddemus** [comment]
Agree with having a label: it is easy to see the labels when [listing the PRs](https://github.com/pytest-dev/pytest/pulls?q=is%3Apr+sort%3Aupdated-desc+is%3Aclosed). 👍

**Pull request merged**